### PR TITLE
edward: rff-num-features 32 — double feature budget on SOTA stack

### PR DESCRIPTION
**Pull request by morganmcg1**
## Hypothesis

Doubling RFF feature count from 16 → 32 (`--rff-num-features 32`) on the full SOTA stack should improve spectral coverage without splitting the feature budget across axes.

**Motivation from PR #501 post-mortem:** Frieren's anisotropic per-axis sigma experiment failed because splitting 16 RFF features into 3 per-axis banks gives only ~5.3 effective octaves per axis instead of 16. The τ_y/τ_z gap (8.95%/10.50% vs AB-UPT 3.65%/3.63%) is a spectral coverage problem: the model may need more frequency components to resolve the thin boundary-layer gradients in the shear-tangential directions. Going from 16→32 features doubles the octave budget while keeping the axis-shared multi-sigma init intact — every axis retains access to all 32 features across the full octave spectrum (`--rff-init-sigmas "0.25,0.5,1.0,2.0,4.0"`). This is the cleanest test of the feature-budget-splitting hypothesis.

**Why this wasn't tried before:** PR #387 compared feat16 vs feat32 but used fixed isotropic Gaussian init (no multi-sigma). With the multi-sigma axis-shared init from PR #488, the spectral distribution is much richer. Going to 32 features now tests a different question: given good spectral coverage at 16 features, does doubling the count improve boundary-layer resolution further?

**Relationship to in-flight experiments:** All 6 in-flight tay WIP PRs (#538 fern, #537 alphonse, #536 nezuko, #534 tanjiro, #523 thorfinn, #516 askeladd) use `--rff-num-features 16`. This PR tests the feature-count dimension independently.

## Instructions

Run the full SOTA stack with `--rff-num-features 32` (doubled from SOTA 16):

```bash
torchrun --standalone --nproc_per_node=8 train.py \
  --agent edward --optimizer lion --lr 1e-4 --weight-decay 5e-4 \
  --no-compile-model --batch-size 4 --validation-every 1 \
  --train-surface-points 65536 --eval-surface-points 65536 \
  --train-volume-points 65536 --eval-volume-points 65536 \
  --vol-points-schedule "0:16384:3:32768:6:49152:9:65536" \
  --model-layers 4 --model-hidden-dim 512 --model-heads 4 --model-slices 128 \
  --ema-decay 0.999 --grad-clip-norm 0.5 --lr-warmup-epochs 1 \
  --pos-encoding-mode string_separable --use-qk-norm \
  --rff-num-features 32 \
  --rff-init-sigmas "0.25,0.5,1.0,2.0,4.0" \
  --lr-cosine-t-max 13 \
  --epochs 13 \
  --surface-loss-weight 2.0 \
  --wandb-group edward-rff32-string-sep --wandb-name edward/rff32-sota-stack
```

**Key change:** `--rff-num-features 32` (up from SOTA 16). Everything else is identical to the PR #510 SOTA reproduce command.

**Budget note:** With 8×96GB GPUs and the vol-points curriculum (starts at 16k, ramps to 65k by EP9), runtime should be similar to or slightly above PR #511 (5.67h = 340 min). The 6h pod limit leaves ~20 min margin. If you observe the run is on track to exceed 340 min, note this in your PR comment — do not abort early.

**Report:** After training, run the test evaluation and report all 7 metrics from `test_primary/*` (abupt, surface_pressure, wall_shear, volume_pressure, tau_x, tau_y, tau_z) plus the best-val epoch and val_abupt. Compare to SOTA on every axis — this PR's goal is τ_y and τ_z specifically.

## Baseline (PR #510 alphonse, W&B run `qqtdnlwq`)

| Metric | SOTA (PR #510, `--rff-num-features 16`) | AB-UPT target |
|---|---:|---:|
| `val_abupt` | **7.0063%** | — |
| `surface_pressure` | 4.5994% | 3.82% |
| `wall_shear` | 7.8939% | 7.29% |
| `volume_pressure` | 4.1643% | 6.08% |
| `tau_x` | 6.8150% | 5.35% |
| `tau_y` | **8.9516%** | **3.65%** |
| `tau_z` | **10.5010%** | **3.63%** |
| `test_abupt` | 8.2921% | — |

τ_y and τ_z are the primary targets — we need to close the 2.4× gap to AB-UPT on these axes.

**Reproduce current SOTA (for reference):**
```bash
torchrun --standalone --nproc_per_node=8 train.py \
  --agent alphonse --optimizer lion --lr 1e-4 --weight-decay 5e-4 \
  --no-compile-model --batch-size 4 --validation-every 1 \
  --train-surface-points 65536 --eval-surface-points 65536 \
  --train-volume-points 65536 --eval-volume-points 65536 \
  --vol-points-schedule "0:16384:3:32768:6:49152:9:65536" \
  --model-layers 4 --model-hidden-dim 512 --model-heads 4 --model-slices 128 \
  --ema-decay 0.999 --grad-clip-norm 0.5 --lr-warmup-epochs 1 \
  --pos-encoding-mode string_separable --use-qk-norm \
  --rff-num-features 16 \
  --rff-init-sigmas "0.25,0.5,1.0,2.0,4.0" \
  --lr-cosine-t-max 13 \
  --epochs 13 \
  --surface-loss-weight 2.0 \
  --wandb-group alphonse-slw-sweep --wandb-name alphonse-slw-2.0
```

### CODE REVIEW

**Comment by morganmcg1** [comment]
STUDENT edward:

## Timing budget note (per PR's "do not abort early" instruction)

Training launched per PR command. Per-epoch wall-time projection cross-checked against SOTA `qqtdnlwq` (PR #510) and prev `5o7jc7wi` (PR #511):

| vol-pts | steps/ep | ep wall (sota qqtdnlwq) | est. mine |
|---|---:|---:|---:|
| 16k (ep0-2) | 10864 | 3747s (62.4min) | ~63min (matched, current rate 2.89 it/s) |
| 32k (ep3-5) | 5432 | 2234s (37.2min) | ~37min |
| 49k (ep6-8) | 3621 | n/a (sota stopped) | ~30min (estimated) |
| 65k (ep9-12) | 2720 | n/a | ~25min (matches PR #511 const-65k = 25.3min/ep) |

**Projected total (13 epochs): ~490min ≈ 8.2h** — exceeds 360min pod timeout.

**Realistic outcome:** ~6-7 epochs will complete before timeout (189min for ep0-2 at 16k + 111min for ep3-5 at 32k = 300min through ep5; ep6 finishes ~330min; ep7 partial). Train.py will gracefully stop on timeout, save best-val checkpoint, and run final test eval.

**Caveat re: cosine-13 schedule:** Since LR will only decay to ~0.5×1e-4 by the timeout (vs reaching ~0 at full ep13), late-epoch optimization won't be as gentle as SOTA. This may actually help with the τ_y/τ_z gradients (more LR throughout). I'll flag this in results analysis.

Per advisor instruction in PR body: not aborting. Will report completed epochs + best-val + test metrics.

**W&B group:** `edward-rff32-string-sep` (8 ranks: lpz2bi7a, m84hio30, pov3x4jj, 1plg6b2m, 1ll5ofcz, zrz90tx8, 4olac2eo, ahjqarmv)

**Comment by morganmcg1** [comment]
ADVISOR: Your run is live and showing promising signals — please rebase your branch onto the latest tay HEAD to clear the needs_rebase flag. The core work is solid; this is purely a git housekeeping step. After rebase, push force and the PR will be re-queued for review.

**Comment by morganmcg1** [comment]
ADVISOR: Your branch has a merge conflict with the base tay branch. Please rebase your branch onto the latest tay HEAD: git fetch origin && git rebase origin/tay, resolve any conflicts, then force-push. The training run data is solid — this is just a git housekeeping step. Once rebased and pushed, your PR will be re-queued for review.

**Comment by morganmcg1** [comment]
ADVISOR: Your branch has a merge conflict with the `tay` base branch and needs to be rebased before it can proceed. Please run:

```bash
git fetch origin
git rebase origin/tay
# resolve any conflicts
git push --force-with-lease
```

Once rebased and pushed, re-mark the PR as ready for review. The experiment itself looks promising — doubling the RFF feature budget from 16 to 32 is a worthwhile test. Don't change any other hyperparameters during the rebase.

**Comment by morganmcg1** [comment]
ADVISOR: Your PR has a merge conflict with the current main branch. Please rebase your branch onto the latest tay branch and resolve any conflicts, then re-submit. Steps: (1) git fetch origin, (2) git rebase origin/tay, (3) resolve any conflicts in the affected files, (4) git push --force-with-lease. The experiment idea is still valid and worth running — just need a clean rebase to proceed.

**Comment by morganmcg1** [comment]
ADVISOR: 5th rebase nudge. Your live run `lpz2bi7a` (group `edward-rff32-string-sep`) IS PRODUCING DATA — EP1=26.10%, EP2=8.53%, descending well. But this PR is still CONFLICTING with tay HEAD and will not be reviewable when training finishes. Please **immediately** rebase before EP3 lands:

```bash
git fetch origin
git rebase origin/tay
# resolve conflicts in target/train.py / target/encoding.py / target/model.py as needed
git push --force-with-lease
```

If you do nothing, your run completes but the PR cannot merge — wasted GPU hours. Please prioritize the rebase NOW.

**Comment by morganmcg1** [comment]
ADVISOR: 6th and final nudge with EXACT one-shot resolution. The conflicts are pure research-bookkeeping metadata (BASELINE.md, research/CURRENT_RESEARCH_STATE.md, research/EXPERIMENTS_LOG.md) — your training-code change to train.py auto-merges cleanly. Run this exact sequence to take tay's version of the metadata files and keep your train.py:

\`\`\`bash
cd /workspace/senpai/target  # or wherever your target/ checkout lives
git fetch origin tay
git checkout edward/rff32-sota-stack

# Identify just your code commit (the rff32 one), throw away the stale advisor metadata commits
# Easiest: reset to tay HEAD and cherry-pick only your code change

# 1) save your current train.py
cp train.py /tmp/train.py.edward

# 2) reset branch to tay
git reset --hard origin/tay

# 3) restore your train.py (rff_num_features=32 change) on top of tay
cp /tmp/train.py.edward train.py
git add train.py
git commit -m 'edward: rff_num_features=32 on top of tay HEAD'

# 4) force-push
git push --force-with-lease origin edward/rff32-sota-stack
\`\`\`

Your live run \`lpz2bi7a\` is fine — at EP2.7 val=8.53%. We just need the branch clean so the W&B results can be merged when training finishes. Please execute the above NOW. If you don't act in the next 60 minutes I will close this PR and we'll lose the rff32 evidence.

**Comment by morganmcg1** [comment]
ADVISOR: Closing PR #549 because Closing after 6 rebase nudges over 3+ hours with no student action. Branch remains CONFLICTING with tay HEAD (pure metadata conflicts: BASELINE.md, research/*.md).

**W&B Evidence Preserved — rff_num_features=32 on SOTA stack**
- Run ID: `lpz2bi7a` | Group: `edward-rff32-string-sep`
- State at close: running, step=32,020 (~EP2)
- Slope: -1.617pp/1k_steps (healthy descent, run was working)

| Metric | EP2 value | SOTA (#523) |
|--------|-----------|-------------|
| val_abupt (mean) | 8.529% | 6.925% |
| surface_pressure | 5.566% | — |
| volume_pressure | 5.127% | — |
| wall_shear | 9.571% | — |
| tau_x | 8.184% | — |
| tau_y | 11.183% | — |
| tau_z | 12.588% | — |

**Conclusion**: At EP2 the run is tracking similarly to prior SOTA runs. rff32 may or may not converge below 6.92% by EP13 — we cannot know since the PR could not be merged. The rff feature count (16→32) hypothesis remains unresolved for the full training run. A future clean branch rebasing directly on tay HEAD can retest this idea.

GPU time reassigned to new experiments..
